Parsimony/ParserBuilder/OptionBuilder.cs

```diff
@@ -0,0 +1,169 @@
+            Selector = new PropertySelector<TOption, TProp>(selector);
+        }
+
+
+        /// <summary>
+        /// Constructs a new option builder with the given short name
+        /// </summary>
+        /// <param name="longName"> a valid long name for the option </param>
+        /// <param name="selector"> The property selector that indicates the property this flag represents on the parent <typeparamref name="TOption"/> </param>
+        /// <exception cref="ArgumentException"> <paramref name="longName"/> is invalid </exception>
+        /// <exception cref="ArgumentException"> <paramref name="selector"/> is not a valid <see cref="PropertySelector{TInput, TProp} "/></exception>
+        /// <exception cref="ArgumentNullException"> <paramref name="longName"/> or <paramref name="selector"/> is null </exception>
+        public OptionBuilder(string longName, Expression<Func<TOption, TProp>> selector)
+        {
+            _longName = OptionName.Parse(longName) as OptionName.Long;
+            Selector = new PropertySelector<TOption, TProp>(selector);
+        }
+
+
+        /// <summary>
+        /// Compiles the rule sets and parser into an option parser of type <see cref="OptionParser{TOptions}"/>
+        /// </summary>
+        /// <exception cref="InvalidOperationException"> Incompatible rules have been found </exception>
+        /// <exception cref="InvalidOperationException"> Duplicate rules are found </exception>
+        /// /// <exception cref="InvalidOperationException">
+        /// No parser is set and no <see cref="TypeConverter"/> can be found that can perform conversions from string 
+        /// </exception>
+        /// <returns> A result containing the parser and the ruleset assocaited with this option </returns>
+        internal OptionParserBuildResult<TOption> Build()
+        {
+            var requiredSet = EnsureNoDuplicateTargets(Rules.Where(x => x.Kind == RuleKind.Requires));
+            var precludesSet = EnsureNoDuplicateTargets(Rules.Where(x => x.Kind == RuleKind.Precludes));
+
+            if (precludesSet.Any(x => requiredSet.Contains(x)))
+                throw new InvalidOperationException("Cannot both require and preclude a property");
+
+            if (Parser == null)
+            {
+                var converter = TypeDescriptor.GetConverter(typeof(TProp));
+                if (!converter.CanConvertFrom(typeof(string)))
+                {
+                    throw new InvalidOperationException(
+                        $"The given type {typeof(TOption).Name} cannot be converted from string," +
+                        $" specify a parser with the extension method 'WithParser'");
+                }
+
+                Parser = (string input) => (TProp)converter.ConvertFromString(input);
+            }
+
+            var opt = new Option<TOption, TProp>(_shortName, _longName, Parser, Selector.Setter);
+
+            return new OptionParserBuildResult<TOption>(opt, Rules);
```

Comment:
I keep thinking the rules should be on the (logical) collection of the options since checks like no intersection between precludes and requires need to be across options. not sure what the implication of that is, whether these just get rolled together later or if changes the interface. ill make an issue later, just want to capture the thought now

Reply:
I think it makes more sense for the builder to pass it up and the parent to handle it. I'm not particularly committed to the idea it just seems like the cleanest from my perspective